README.md
# rucksack

Texture packer and resource bundler. Use the rucksack executable to build your
resources file, and then depend on librucksack in-game to get the resources out
at runtime. Alternately, roll your own resource-loading code using the file
format detailed at the end of this README.

rucksack defines a resource manifest file format which is loosely based on JSON
but accepts comments, extra punctuation, and unquoted strings. This manifest
file tells rucksack which files to bundle and how to store them. rucksack uses
a rectangle bin packing algorithm to efficiently store multiple images into a
single texture, otherwise known as a "sprite sheet".

## Command Line Usage

```
rucksack v3.1.0

Usage: ./rucksack [command] [command-options]

Commands:
  help       get info on how to use a command
  bundle     parses an assets json file and keeps a bundle up to date
  cat        extracts a single file from the bundle and writes it to stdout
  ls         lists all resources in a bundle
  rm         remove a file from the bundle
  strip      make an existing bundle as small as possible
  unpack     create a directory with the bundle contents
```

## Library Usage

```C
#include <rucksack/rucksack.h>

int main() {
    struct RuckSackBundle *bundle;
    rucksack_bundle_open(bundle_name, &bundle);

    struct RuckSackFileEntry *entry = rucksack_bundle_find_file(bundle, "blah");
    size_t size = rucksack_file_size(entry);
    unsigned char *buffer = malloc(size);
    rucksack_bundle_file_read(bundle, entry, buf);
    // buffer now contains the contents of the file indexed by "blah"
    free(buffer);
    rucksack_bundle_close(bundle);
}
```

## Dependencies

 * [FreeImage](http://freeimage.sourceforge.net/)
 * [liblaxjson](https://github.com/andrewrk/liblaxjson)

## Installation

### Pre-Built Packages

 * [Ubuntu PPA](https://launchpad.net/~andrewrk/+archive/rucksack)

   ```
   sudo apt-add-repository ppa:andrewrk/rucksack
   sudo apt-get update
   sudo apt-get install rucksack
   ```

### From Source

1. `mkdir build && cd build && cmake ..`
2. Verify that all dependencies say "OK".
3. `make && sudo make install`

## JSON Reference

### Example

```js
// comments are OK :)
// single quotes, double quotes, and no quotes are OK
{
  // one-off files you want to directly save into the bundle
  files: {
    file1Name: {
      path: "path/to/file",
    },
  },
  // if you want to avoid manually specifying every file, you can glob
  globFiles: [
    {
      path: "path/to/dir",
      glob: "*",
      prefix: "abc_", // prepended to the key
    },
  ],
  // spritesheet generation
  textures: {
    texture1Name: {
      // 1024x1024 is the default.
      maxWidth: 1024,
      maxHeight: 1024,

      // By default rucksack enforces that the width and height of a texture
      // will be a power of 2. Set pow2 to false to disable this.
      pow2: true,

      // rucksack efficiently packs textures by sometimes rotating images
      // to make them fit better, and you can compensate for this by using
      // different UV coords when the images are rotated. If you for some
      // reason want to disable this feature, you can set allowRotate90 to
      // false.
      allowRotate90: true,

      globImages: [
        {
          path: "path/to/dir",
          glob: "*",
          prefix: "abc_", // prepended to the key

          // can be any of these strings:
          // "top", "left", "bottom", "right"
          // "topleft", "topright", "bottomleft", "bottomright"
          // "center"
          // or it can be an object like this: {x: 13, y: 19}
          anchor: "center",
        },
      ],
      images: {
        image1Name: {
          // can be any image format supported by FreeImage
          path: "path/to/image.png",

          // can be any of these strings:
          // "top", "left", "bottom", "right"
          // "topleft", "topright", "bottomleft", "bottomright"
          // "center"
          // or it can be an object like this: {x: 13, y: 19}
          anchor: "center",
        },
      },
    },
  },
}
```

## RuckSack Bundle File Format

The main header identifies the file and tells you some metadata about the
rest of the header entries.

The basic data supported by rucksack is a file. In this case the offset that
a header entry points to is the raw file data. Some files in a rucksack bundle
are textures, also known as spritesheets. This is described by Texture Format
below.

    Offset | Contents
    -------+---------
         0 | 16 byte UUID - 60 70 c8 99 82 a1 41 84 89 51 08 c9 1c c9 b6 20
        16 | uint32be file format version. bumped when incompatible changes made.
        20 | uint32be offset of first header entry from file start
        24 | uint32be number of header entries

### Header Entry Format

    Offset | Contents
    -------+---------
         0 | uint32be size of this header entry in bytes
         4 | uint64be offset of this entry's file contents
        12 | uint64be actual size of the file contents in bytes
        20 | uint64be number of allocated bytes for this file
        28 | uint32be file mtime
        32 | uint32be key size in bytes
        36 | key bytes

### Texture Format

    Offset | Contents
    -------+---------
         0 | 16 byte UUID 0e b1 4c 84 47 4c b3 ad a6 bd 93 e4 be a5 46 ba
        16 | uint32be offset of the actual image data from 0 in this struct
        20 | uint32be number of images in this texture
        24 | uint32be offset of the first image entry from 0 in this struct
        28 | uint32be max_width used when creating this texture
        32 | uint32be max_height used when creating this texture
        36 | uint8 pow2 value used when creating this texture
        37 | uint8 allow_r90 value used when creating this texture

#### Image Entry Format

    Offset | Contents
    -------+---------
         0 | uint32be size of this image entry header in bytes
         4 | uint32be anchor position enum value
         8 | uint32be anchor x multiplied by 16384
        12 | uint32be anchor y multiplied by 16384
        16 | uint32be image x
        20 | uint32be image y
        24 | uint32be unrotated image width
        28 | uint32be unrotated image height
        32 | uint8 boolean whether the image is rotated clockwise 90 degrees
        33 | uint32be key size in bytes
        37 | key bytes

## Projects Using rucksack

Feel free to make a pull request adding to this list.

 * [Legend of Swarkland](https://github.com/thejoshwolfe/legend-of-swarkland) - Basic turn-based, hack-n-slash survival game.
 * [spacefight](https://github.com/andrewrk/spacefight) - 3D asteroids prototype
 * [grapple](https://github.com/andrewrk/grapple) - prototyping a 4 player arcade game based on a grapple gun and a physics engine


CHANGELOG.md
### 3.1.0

 * `bundle` command has a `--force-r90` flag to make it easier to test code to
   handle images rotated 90 degrees.
 * add `rucksack_bundle_open_read_mem` API for reading a rucksack bundle from
   memory.

### 3.0.2

 * build: use mkdir in a way that works with mingw
 * build: detect if glob support is available
 * build: add missing library dependency

### 3.0.1

 * librucksack supports deleting entries
 * add `rm` command to delete entries
 * librucksack is now two libraries: librucksack and librucksackspritesheet.
   librucksack does not depend on FreeImage. Users will typically only need
   to link against librucksack.
 * add `unpack` command
 * `bundle` command deletes entries no longer found in manifest file

### 3.0.0

(redacted)

### 2.3.2

 * add `#define NDEBUG` to tests to make tests build more portably

### 2.3.1

 * avoid `strdup` and `strndup` for better portability

### 2.3.0

 * fix incorrect error value for read-only file not found
 * fix errors leaving bundle file in invalid state
 * `--dep` feature: globbing adds dependency on directory timestamp
 * support path lengths up to 4096 bytes
 * add API to get the length of an entry key: `rucksack_file_name_size`
 * fix joining an empty path with the current working directory
 * add `strip` command to make a bundle as small as possible.
 * fix several bugs that could result in an invalid bundle

### 2.2.0

 * ability to open bundle read-only
 * read-only commands do not create empty bundles when files do not exist

### 2.1.1

 * fix segfault when not using `--deps` option

### 2.1.0

 * `bundle` command takes a new parameter `--deps` to create a .d
   (makefile format) dependencies file

### 2.0.1

 * fix incorrect fixed point length (16384 instead of 10000)

### 2.0.0

 * ability to open a rucksack bundle which is a 0 byte empty file.
 * added a setting which allows preventing rotating images 90 degrees.
 * ability to glob for images in a texture.
 * fix not updating bundle when texture properties or image anchor position
   changed in manifest json
 * fix 2 bugs causing crash when adding or updating file in bundle
 * cat command: generate JSON data for textures
 * library: fix explicit anchor using bogus values
 * bundle command: fix explicit anchor using bogus values
 * use width/height for anchor position right/bottom instead of width-1/height-1
 * (API breaking) rename all `page` functions to `texture`.
 * (API breaking) rucksack_texture_close removed. Use rucksack_texture_destroy
   instead.
 * (API breaking) remove `RuckSackImage` property `name` in favor of `key`,
   `key_size`, and `path`.
 * (format breaking) update bundle format
   - include bundle version number
   - texture header includes UUID to identify it as a texture
   - texture header includes the properties that were used to generate it
   - image header includes anchor position
   - image header uses only 1 byte for the r90 boolean
 * (API breaking) add rucksack_image_create and rucksack_image_destroy.
   RuckSackImage struct size no longer part of ABI.
   - now you can set image->key and ignore image->key_size if you have a null-
     terminated string.
 * (API breaking) key and key_size are passed in via RuckSackTexture struct
   instead of separate arguments to rucksack_file_open_texture
 * (API breaking) all API that deals with keys now has a key_size parameter. All
   keys are arrays of bytes with a length, and keys may have null bytes in them.
 * (API breaking) anchor_x and anchor_y are floats instead of ints

### 1.0.1

 * Update docs and add changelog

### 1.0.0

 * Initial release.


src/laxjson.c
/*
 * Copyright (c) 2013 Andrew Kelley
 *
 * This file is part of rucksack, which is MIT licensed.
 * See http://opensource.org/licenses/MIT
 */

#include "laxjson.h"

#include <stdlib.h>
#include <assert.h>

#define WHITESPACE \
    ' ': \
    case '\t': \
    case '\n': \
    case '\f': \
    case '\r': \
    case 0xb

#define DIGIT \
    '0': \
    case '1': \
    case '2': \
    case '3': \
    case '4': \
    case '5': \
    case '6': \
    case '7': \
    case '8': \
    case '9'

#define ALPHANUMERIC \
    'a': \
    case 'b': \
    case 'c': \
    case 'd': \
    case 'e': \
    case 'f': \
    case 'g': \
    case 'h': \
    case 'i': \
    case 'j': \
    case 'k': \
    case 'l': \
    case 'm': \
    case 'n': \
    case 'o': \
    case 'p': \
    case 'q': \
    case 'r': \
    case 's': \
    case 't': \
    case 'u': \
    case 'v': \
    case 'w': \
    case 'x': \
    case 'y': \
    case 'z': \
    case 'A': \
    case 'B': \
    case 'C': \
    case 'D': \
    case 'E': \
    case 'F': \
    case 'G': \
    case 'H': \
    case 'I': \
    case 'J': \
    case 'K': \
    case 'L': \
    case 'M': \
    case 'N': \
    case 'O': \
    case 'P': \
    case 'Q': \
    case 'R': \
    case 'S': \
    case 'T': \
    case 'U': \
    case 'V': \
    case 'W': \
    case 'X': \
    case 'Y': \
    case 'Z': \
    case DIGIT

#define VALID_UNQUOTED \
    '-': \
    case '_': \
    case '#': \
    case '$': \
    case '%': \
    case '&': \
    case '<': \
    case '>': \
    case '=': \
    case '~': \
    case '|': \
    case '@': \
    case '?': \
    case ';': \
    case '.': \
    case '+': \
    case '*': \
    case '(': \
    case ')'
    

static const int HEX_MULT[] = {4096, 256, 16, 1};

static enum LaxJsonError push_state(struct LaxJsonContext *context, enum LaxJsonState state) {
    if (context->state_stack_index >= context->state_stack_size) {
        context->state_stack_size += 1024;
        if (context->state_stack_size > context->max_state_stack_size)
            return LaxJsonErrorExceededMaxStack;
        enum LaxJsonState *new_ptr = realloc(context->state_stack, context->state_stack_size);
        if (!new_ptr)
            return LaxJsonErrorNoMem;
        context->state_stack = new_ptr;
    }
    context->state_stack[context->state_stack_index] = state;
    context->state_stack_index += 1;
    return LaxJsonErrorNone;
}

struct LaxJsonContext *lax_json_create(void) {
    struct LaxJsonContext *context = calloc(1, sizeof(struct LaxJsonContext));

    context->value_buffer_size = 1024;
    context->value_buffer = malloc(context->value_buffer_size);
    // TODO handle malloc returning NULL
    if (!context->value_buffer) abort();

    context->state_stack_size = 1024;
    context->state_stack = malloc(context->state_stack_size);
    if (!context->state_stack) abort();

    context->line = 1;
    context->max_state_stack_size = 16384;
    context->max_value_buffer_size = 1048576; // 1 MB

    push_state(context, LaxJsonStateEnd);

    return context;
}

void lax_json_destroy(struct LaxJsonContext *context) {
    free(context->state_stack);
    free(context->value_buffer);
    free(context);
}

static void pop_state(struct LaxJsonContext *context) {
    context->state = context->state_stack[context->state_stack_index];
    context->state_stack_index -= 1;
    assert(context->state_stack_index >= 0);
}

static enum LaxJsonError buffer_char(struct LaxJsonContext *context, char c) {
    if (context->value_buffer_index >= context->value_buffer_size) {
        context->value_buffer_size += 16384;
        if (context->value_buffer_size > context->max_value_buffer_size)
            return LaxJsonErrorExceededMaxValueSize;
        char *new_ptr = realloc(context->value_buffer, context->value_buffer_size);
        if (!new_ptr)
            return LaxJsonErrorNoMem;
        context->value_buffer = new_ptr;
    }
    context->value_buffer[context->value_buffer_index] = c;
    context->value_buffer_index += 1;
    return LaxJsonErrorNone;
}

enum LaxJsonError lax_json_feed(struct LaxJsonContext *context, int size, char *data) {
#define PUSH_STATE(state) \
    err = push_state(context, state); \
    if (err) return err;
#define BUFFER_CHAR(c) \
    err = buffer_char(context, c); \
    if (err) return err;

    enum LaxJsonError err = LaxJsonErrorNone;
    int x;
    for (char *end = data + size; data < end; data += 1) {
        char c = *data;
        if (c == '\n') {
            context->line += 1;
            context->column = 1;
        } else {
            context->column += 1;
        }
        switch (context->state) {
            case LaxJsonStateEnd:
                switch (c) {
                    case WHITESPACE:
                        // ignore
                        break;
                    case '/':
                        context->state = LaxJsonStateCommentBegin;
                        PUSH_STATE(LaxJsonStateEnd);
                        break;
                    default:
                        return LaxJsonErrorExpectedEof;
                }
                break;
            case LaxJsonStateObject:
                switch (c) {
                    case WHITESPACE:
                    case ',':
                        // do nothing except eat these characters
                        break;
                    case '/':
                        context->state = LaxJsonStateCommentBegin;
                        PUSH_STATE(LaxJsonStateObject);
                        break;
                    case '"':
                    case '\'':
                        context->state = LaxJsonStateString;
                        context->value_buffer_index = 0;
                        context->delim = c;
                        context->string_type = LaxJsonTypeProperty;
                        PUSH_STATE(LaxJsonStateColon);
                        break;
                    case VALID_UNQUOTED:
                        context->state = LaxJsonStateBareProp;
                        context->value_buffer[0] = c;
                        context->value_buffer_index = 1;
                        context->delim = 0;
                        break;
                    case '}':
                        context->end(context, LaxJsonTypeObject);
                        pop_state(context);
                        break;
                    default:
                        return LaxJsonErrorUnexpectedChar;
                }
                break;
            case LaxJsonStateBareProp:
                switch (c) {
                    case VALID_UNQUOTED:
                        BUFFER_CHAR(c);
                        break;
                    case WHITESPACE:
                        BUFFER_CHAR('\0');
                        context->string(context, LaxJsonTypeProperty, context->value_buffer);
                        context->state = LaxJsonStateColon;
                        break;
                    case ':':
                        BUFFER_CHAR('\0');
                        context->string(context, LaxJsonTypeProperty, context->value_buffer);
                        context->state = LaxJsonStateValue;
                        context->string_type = LaxJsonTypeString;
                        PUSH_STATE(LaxJsonStateObject);
                        break;
                    default:
                        return LaxJsonErrorUnexpectedChar;
                }
                break;
            case LaxJsonStateString:
                if (c == context->delim) {
                    BUFFER_CHAR('\0');
                    context->string(context, context->string_type, context->value_buffer);
                    pop_state(context);
                } else if (c == '\\') {
                    context->state = LaxJsonStateStringEscape;
                } else {
                    BUFFER_CHAR(c);
                }
                break;
            case LaxJsonStateStringEscape:
                switch (c) {
                    case '\'':
                    case '"':
                    case '/':
                    case '\\':
                        BUFFER_CHAR(c);
                        break;
                    case 'b':
                        BUFFER_CHAR('\b');
                        break;
                    case 'f':
                        BUFFER_CHAR('\f');
                        break;
                    case 'n':
                        BUFFER_CHAR('\n');
                        break;
                    case 'r':
                        BUFFER_CHAR('\r');
                        break;
                    case 't':
                        BUFFER_CHAR('\t');
                        break;
                    case 'u':
                        context->state = LaxJsonStateUnicodeEscape;
                        context->unicode_digit_index = 0;
                        context->unicode_point = 0;
                        break;
                }
                break;
            case LaxJsonStateUnicodeEscape:
                switch (c) {
                    case '0':
                        x = 0;
                        break;
                    case '1':
                        x = 1;
                        break;
                    case '2':
                        x = 2;
                        break;
                    case '3':
                        x = 3;
                        break;
                    case '4':
                        x = 4;
                        break;
                    case '5':
                        x = 5;
                        break;
                    case '6':
                        x = 6;
                        break;
                    case '7':
                        x = 7;
                        break;
                    case '8':
                        x = 8;
                        break;
                    case '9':
                        x = 9;
                        break;
                    case 'a':
                    case 'A':
                        x = 10;
                        break;
                    case 'b':
                    case 'B':
                        x = 11;
                        break;
                    case 'c':
                    case 'C':
                        x = 12;
                        break;
                    case 'd':
                    case 'D':
                        x = 13;
                        break;
                    case 'e':
                    case 'E':
                        x = 14;
                        break;
                    case 'f':
                    case 'F':
                        x = 15;
                        break;
                    default:
                        return LaxJsonErrorInvalidHexDigit;
                }
                context->unicode_point += x * HEX_MULT[context->unicode_digit_index];
                context->unicode_digit_index += 1;
                if (context->unicode_digit_index == 4) {
                    if (context->unicode_point <= 0x007f) {
                        // 1 byte ascii
                        BUFFER_CHAR((char)context->unicode_point);
                        context->state = LaxJsonStateString;
                    } else if (context->unicode_point <= 0x07ff) {
                        // 2 bytes
                        // TODO
                        abort();
                    } else if (context->unicode_point <= 0xffff) {
                        // 3 bytes
                        // TODO
                        abort();
                    } else if (context->unicode_point <= 0x1fffff) {
                        // 4 bytes
                        // TODO
                        abort();
                    } else if (context->unicode_point <= 0x3ffffff) {
                        // 5 bytes
                        // TODO
                        abort();
                    } else if (context->unicode_point <= 0x7fffffff) {
                        // 6 bytes
                        // TODO
                        abort();
                    } else {
                        return LaxJsonErrorInvalidUnicodePoint;
                    }
                }
                break;
            case LaxJsonStateColon:
                switch (c) {
                    case WHITESPACE:
                        // ignore it
                        break;
                    case '/':
                        context->state = LaxJsonStateCommentBegin;
                        PUSH_STATE(LaxJsonStateColon);
                        break;
                    case ':':
                        context->state = LaxJsonStateValue;
                        context->string_type = LaxJsonTypeString;
                        PUSH_STATE(LaxJsonStateObject);
                        break;
                    default:
                        return LaxJsonErrorExpectedColon;
                }
                break;
            case LaxJsonStateValue:
                switch (c) {
                    case WHITESPACE:
                        // ignore
                        break;
                    case '/':
                        context->state = LaxJsonStateCommentBegin;
                        PUSH_STATE(LaxJsonStateValue);
                        break;
                    case '{':
                        context->begin(context, LaxJsonTypeObject);
                        context->state = LaxJsonStateObject;
                        break;
                    case '[':
                        context->begin(context, LaxJsonTypeArray);
                        context->state = LaxJsonStateArray;
                        break;
                    case '\'':
                    case '"':
                        context->state = LaxJsonStateString;
                        context->delim = c;
                        context->value_buffer_index = 0;
                        break;
                    case '-':
                        context->state = LaxJsonStateNumber;
                        context->value_buffer[0] = c;
                        context->value_buffer_index = 1;
                        break;
                    case '+':
                        context->state = LaxJsonStateNumber;
                        context->value_buffer_index = 0;
                        break;
                    case DIGIT:
                        context->state = LaxJsonStateNumber;
                        context->value_buffer_index = 1;
                        context->value_buffer[0] = c;
                        break;
                    case 't':
                        context->primitive(context, LaxJsonTypeTrue);
                        context->state = LaxJsonStateExpect;
                        context->expected = "rue";
                        break;
                    case 'f':
                        context->primitive(context, LaxJsonTypeFalse);
                        context->state = LaxJsonStateExpect;
                        context->expected = "alse";
                        break;
                    case 'n':
                        context->primitive(context, LaxJsonTypeNull);
                        context->state = LaxJsonStateExpect;
                        context->expected = "ull";
                        break;
                    default:
                        return LaxJsonErrorUnexpectedChar;
                }
                break;
            case LaxJsonStateArray:
                switch (c) {
                    case WHITESPACE:
                    case ',':
                        // ignore
                        break;
                    case '/':
                        context->state = LaxJsonStateCommentBegin;
                        PUSH_STATE(LaxJsonStateArray);
                        break;
                    case ']':
                        pop_state(context);
                        break;
                    default:
                        context->state = LaxJsonStateValue;
                        PUSH_STATE(LaxJsonStateArray);

                        // rewind 1 character
                        data -= 1;
                        context->column -= 1;
                        continue;
                }
                break;
            case LaxJsonStateNumber:
                switch (c) {
                    case DIGIT:
                        BUFFER_CHAR(c);
                        break;
                    case '.':
                        BUFFER_CHAR(c);
                        context->state = LaxJsonStateNumberDecimal;
                        break;
                    default:
                        return LaxJsonErrorUnexpectedChar;
                }
                break;
            case LaxJsonStateNumberDecimal:
                switch (c) {
                    case DIGIT:
                        BUFFER_CHAR(c);
                        break;
                    case 'e':
                    case 'E':
                        BUFFER_CHAR('e');
                        context->state = LaxJsonStateNumberExponentSign;
                        break;
                    default:
                        return LaxJsonErrorUnexpectedChar;
                }
                break;
            case LaxJsonStateNumberExponentSign:
                switch (c) {
                    case '+':
                    case '-':
                        BUFFER_CHAR(c);
                        context->state = LaxJsonStateNumberExponent;
                        break;
                    default:
                        return LaxJsonErrorUnexpectedChar;
                }
            case LaxJsonStateNumberExponent:
                switch (c) {
                    case DIGIT:
                        BUFFER_CHAR(c);
                        break;
                    default:
                        BUFFER_CHAR('\0');
                        context->number(context, atof(context->value_buffer));
                        pop_state(context);

                        // rewind 1
                        data -= 1;
                        context->column -= 1;
                        continue;
                }
            case LaxJsonStateExpect:
                if (c == *context->expected) {
                    context->expected += 1;
                    if (*context->expected == 0) {
                        pop_state(context);
                    }
                } else {
                    return LaxJsonErrorUnexpectedChar;
                }
                break;
            case LaxJsonStateCommentBegin:
                switch (c) {
                    case '/':
                        context->state = LaxJsonStateCommentLine;
                        break;
                    case '*':
                        context->state = LaxJsonStateCommentMultiLine;
                        break;
                    default:
                        return LaxJsonErrorUnexpectedChar;
                }
                break;
            case LaxJsonStateCommentLine:
                if (c == '\n')
                    pop_state(context);
                break;
            case LaxJsonStateCommentMultiLine:
                if (c == '*')
                    context->state = LaxJsonStateCommentMultiLineStar;
                break;
            case LaxJsonStateCommentMultiLineStar:
                if (c == '/')
                    pop_state(context);
                else
                    context->state = LaxJsonStateCommentMultiLine;
                break;
        }
    }
    return err;
}
